```
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/DataFormats/MuonReco/src/Muon.cc: In member function 'unsigned int reco::Muon::stationMask(reco::Muon::ArbitrationType) const':
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/DataFormats/MuonReco/src/Muon.cc:152:18: warning: unused variable 'rpcMatch' [-Wunused-variable]
        for (auto& rpcMatch : chamberMatch.rpcMatches) {
                  ^~~~~~~~
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/DataFormats/MuonReco/src/Muon.cc: In member function 'unsigned int reco::Muon::RPClayerMask(reco::Muon::ArbitrationType) const':
  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-14-1100/src/DataFormats/MuonReco/src/Muon.cc:237:16: warning: unused variable 'rpcMatch' [-Wunused-variable]
      for (auto& rpcMatch : chamberMatch.rpcMatches) {  // There is clearly something odd here
                ^~~~~~~~
```

Indeed, this is exactly the compilation warning that allowed me to pinpoint the issue reported in https://github.com/cms-sw/cmssw/issues/31783

I can apply a patch here to reproduce the exact previous behavior without getting that warning (" just check if the size of `chamberMatch->rpcMatches` is above 0, compute once `curmask` in that case, and add it to the `totMask`").

But maybe it is better to understand the original intentions of the author of the code and apply a more appropriate fix, if needed

--- a/DataFormats/MuonReco/interface/Muon.h
+++ b/DataFormats/MuonReco/interface/Muon.h
@@ -419,12 +419,14 @@ namespace reco {
 
     float t0(int n = 0) {
       int i = 0;
-      for (auto& chamber : muMatches_)
-        for (auto& segment : chamber.segmentMatches) {
-          if (i == n)
-            return segment.t0;
-          ++i;
+      for (auto& chamber : muMatches_) {
+        int segmentMatchesSize = (int)chamber.segmentMatches.size();
+        if (i + segmentMatchesSize < n) {
+          i += segmentMatchesSize;
+          continue;
         }
+        return chamber.segmentMatches[n - i].t0;
+      }
       return 0;
     }
   };

--- a/DataFormats/MuonReco/src/Muon.cc
+++ b/DataFormats/MuonReco/src/Muon.cc
@@ -149,7 +149,10 @@ unsigned int Muon::stationMask(ArbitrationType type) const {
       RPCDetId rollId = chamberMatch.id.rawId();
       int rpcIndex = rollId.region() == 0 ? 1 : 2;
 
-      for (auto& rpcMatch : chamberMatch.rpcMatches) {
+      //    for (auto& rpcMatch : chamberMatch.rpcMatches) { // TO BE FIXED: there is clearly something odd here
+      // Replaced by something which restores the previous functionality, but one should verify which were the
+      // Original intentions of the author and provide a more appropriate fix (and remove these comment lines)
+      if (!chamberMatch.rpcMatches.empty()) {
         curMask = 1 << ((chamberMatch.station() - 1) + 4 * (rpcIndex - 1));
 
         // do not double count
@@ -234,7 +237,10 @@ unsigned int Muon::RPClayerMask(ArbitrationType type) const {
     } else
       rpcLayer += 6;
 
-    for (auto& rpcMatch : chamberMatch.rpcMatches) {  // There is clearly something odd here
+    //    for (auto& rpcMatch : chamberMatch.rpcMatches) { // TO BE FIXED: there is clearly something odd here
+    // Replaced by something which restores the previous functionality, but one should verify which were the
+    // Original intentions of the author and provide a more appropriate fix (and remove these comment lines)
+    if (!chamberMatch.rpcMatches.empty()) {
       curMask = 1 << (rpcLayer - 1);
 
       // do not double count
@@ -293,8 +299,8 @@ unsigned int Muon::stationGapMaskPull(float sigmaCut) const {
         std::abs(edgeY / yErr) > sigmaCut)  // inside the chamber so negates all gaps for this station
       continue;
 
-    if ((std::abs(edgeX / xErr) < sigmaCut && edgeY / yErr < sigmaCut) ||
-        (std::abs(edgeY / yErr) < sigmaCut && edgeX / xErr < sigmaCut))  // inside gap
+    if ((std::abs(edgeX / xErr) < sigmaCut && edgeY < sigmaCut * yErr) ||
+        (std::abs(edgeY / yErr) < sigmaCut && edgeX < sigmaCut * xErr))  // inside gap
       curMask = 1 << ((stationIndex - 1) + 4 * (detectorIndex - 1));
     totMask += curMask;  // add to total mask
   }
@@ -503,7 +509,7 @@ float Muon::pullX(int station, int muonSubdetId, ArbitrationType type, bool incl
     return 999999;
   if (includeSegmentError)
     return (chamberSegmentPair.first->x - chamberSegmentPair.second->x) /
-           sqrt(pow(chamberSegmentPair.first->xErr, 2) + pow(chamberSegmentPair.second->xErr, 2));
+           sqrt(std::pow(chamberSegmentPair.first->xErr, 2) + std::pow(chamberSegmentPair.second->xErr, 2));
   return (chamberSegmentPair.first->x - chamberSegmentPair.second->x) / chamberSegmentPair.first->xErr;
 }
 
@@ -518,7 +524,7 @@ float Muon::pullY(int station, int muonSubdetId, ArbitrationType type, bool incl
     return 999999;
   if (includeSegmentError)
     return (chamberSegmentPair.first->y - chamberSegmentPair.second->y) /
-           sqrt(pow(chamberSegmentPair.first->yErr, 2) + pow(chamberSegmentPair.second->yErr, 2));
+           sqrt(std::pow(chamberSegmentPair.first->yErr, 2) + std::pow(chamberSegmentPair.second->yErr, 2));
   return (chamberSegmentPair.first->y - chamberSegmentPair.second->y) / chamberSegmentPair.first->yErr;
 }
 
@@ -531,7 +537,7 @@ float Muon::pullDxDz(int station, int muonSubdetId, ArbitrationType type, bool i
     return 999999;
   if (includeSegmentError)
     return (chamberSegmentPair.first->dXdZ - chamberSegmentPair.second->dXdZ) /
-           sqrt(pow(chamberSegmentPair.first->dXdZErr, 2) + pow(chamberSegmentPair.second->dXdZErr, 2));
+           sqrt(std::pow(chamberSegmentPair.first->dXdZErr, 2) + std::pow(chamberSegmentPair.second->dXdZErr, 2));
   return (chamberSegmentPair.first->dXdZ - chamberSegmentPair.second->dXdZ) / chamberSegmentPair.first->dXdZErr;
 }
 
@@ -546,7 +552,7 @@ float Muon::pullDyDz(int station, int muonSubdetId, ArbitrationType type, bool i
     return 999999;
   if (includeSegmentError)
     return (chamberSegmentPair.first->dYdZ - chamberSegmentPair.second->dYdZ) /
-           sqrt(pow(chamberSegmentPair.first->dYdZErr, 2) + pow(chamberSegmentPair.second->dYdZErr, 2));
+           sqrt(std::pow(chamberSegmentPair.first->dYdZErr, 2) + std::pow(chamberSegmentPair.second->dYdZErr, 2));
   return (chamberSegmentPair.first->dYdZ - chamberSegmentPair.second->dYdZ) / chamberSegmentPair.first->dYdZErr;
 }
 

--- a/DataFormats/MuonReco/src/MuonSelectors.cc
+++ b/DataFormats/MuonReco/src/MuonSelectors.cc
@@ -47,7 +47,7 @@ unsigned int muon::RequiredStationMask(const reco::Muon& muon,
     for (int detectorIdx = 1; detectorIdx < 3; ++detectorIdx) {
       float dist = muon.trackDist(stationIdx, detectorIdx, arbitrationType);
       if (dist < maxChamberDist &&
-          dist / muon.trackDistErr(stationIdx, detectorIdx, arbitrationType) < maxChamberDistPull)
+          dist < maxChamberDistPull * muon.trackDistErr(stationIdx, detectorIdx, arbitrationType))
         theMask += 1 << ((stationIdx - 1) + 4 * (detectorIdx - 1));
     }
   }
@@ -558,7 +558,7 @@ bool muon::isGoodMuon(const reco::Muon& muon,
       for (const auto& rpcMatch : chamberMatch.rpcMatches) {
         const float rpcX = rpcMatch.x;
         const float dX = std::abs(rpcX - trkX);
-        if (dX < maxAbsDx or dX / errX < maxAbsPullX) {
+        if (dX < maxAbsDx or dX < maxAbsPullX * errX) {
           ++nMatch;
           break;
         }
@@ -593,10 +593,11 @@ bool muon::isGoodMuon(const reco::Muon& muon,
 
         const float dX = std::abs(me0X - trkX);
         const float dY = std::abs(me0Y - trkY);
-        const float pullX = dX / std::sqrt(errX2 + me0ErrX2);
-        const float pullY = dY / std::sqrt(errY2 + me0ErrY2);
+        const float invPullX2 = errX2 + me0ErrX2;
+        const float invPullY2 = errY2 + me0ErrY2;
 
-        if ((dX < maxAbsDx or pullX < maxAbsPullX) and (dY < maxAbsDy or pullY < maxAbsPullY)) {
+        if ((dX < maxAbsDx or dX < maxAbsPullX * std::sqrt(invPullX2)) and
+            (dY < maxAbsDy or dY < maxAbsPullY * std::sqrt(invPullY2))) {
           ++nMatch;
           break;
         }
@@ -628,10 +629,11 @@ bool muon::isGoodMuon(const reco::Muon& muon,
 
         const float dX = std::abs(gemX - trkX);
         const float dY = std::abs(gemY - trkY);
-        const float pullX = dX / std::sqrt(errX2 + gemErrX2);
-        const float pullY = dY / std::sqrt(errY2 + gemErrY2);
+        const float invPullX2 = errX2 + gemErrX2;
+        const float invPullY2 = errY2 + gemErrY2;
 
-        if ((dX < maxAbsDx or pullX < maxAbsPullX) and (dY < maxAbsDy or pullY < maxAbsPullY)) {
+        if ((dX < maxAbsDx or dX < maxAbsPullX * std::sqrt(invPullX2)) and
+            (dY < maxAbsDy or dY < maxAbsPullY * std::sqrt(invPullY2))) {
           ++nMatch;
           break;
         }
@@ -965,7 +967,7 @@ bool muon::isTrackerHighPtMuon(const reco::Muon& muon, const reco::Vertex& vtx) 
   bool hits = muon.innerTrack()->hitPattern().trackerLayersWithMeasurement() > 5 &&
               muon.innerTrack()->hitPattern().numberOfValidPixelHits() > 0;
 
-  bool momQuality = muon.tunePMuonBestTrack()->ptError() / muon.tunePMuonBestTrack()->pt() < 0.3;
+  bool momQuality = muon.tunePMuonBestTrack()->ptError() < 0.3 * muon.tunePMuonBestTrack()->pt();
 
   bool ip =
       std::abs(muon.innerTrack()->dxy(vtx.position())) < 0.2 && std::abs(muon.innerTrack()->dz(vtx.position())) < 0.5;

--- a/RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc
+++ b/RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc
@@ -49,9 +49,11 @@ MuonIdProducer::MuonIdProducer(const edm::ParameterSet& iConfig)
   addExtraSoftMuons_ = iConfig.getParameter<bool>("addExtraSoftMuons");
   maxAbsEta_ = iConfig.getParameter<double>("maxAbsEta");
   maxAbsDx_ = iConfig.getParameter<double>("maxAbsDx");
-  maxAbsPullX_ = iConfig.getParameter<double>("maxAbsPullX");
+  maxAbsPullX2_ = iConfig.getParameter<double>("maxAbsPullX");
+  maxAbsPullX2_ *= maxAbsPullX2_;
   maxAbsDy_ = iConfig.getParameter<double>("maxAbsDy");
-  maxAbsPullY_ = iConfig.getParameter<double>("maxAbsPullY");
+  maxAbsPullY2_ = iConfig.getParameter<double>("maxAbsPullY");
+  maxAbsPullY2_ *= maxAbsPullY2_;
   fillCaloCompatibility_ = iConfig.getParameter<bool>("fillCaloCompatibility");
   fillEnergy_ = iConfig.getParameter<bool>("fillEnergy");
   storeCrossedHcalRecHits_ = iConfig.getParameter<bool>("storeCrossedHcalRecHits");
@@ -564,7 +566,7 @@ void MuonIdProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) 
         for (auto& muon : *outputMuons) {
           if (muon.innerTrack().get() == trackerMuon.innerTrack().get() &&
               std::abs(reco::deltaPhi(phiOfMuonInteractionRegion(muon), phiOfMuonInteractionRegion(trackerMuon))) <
-                  0.5 * M_PI) {
+                  M_PI_2) {
             newMuon = false;
             muon.setMatches(trackerMuon.matches());
             if (trackerMuon.isTimeValid())
@@ -665,7 +667,7 @@ void MuonIdProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) 
       // predict direction based on the muon interaction region location
       // if it's available
       if (muon.isStandAloneMuon()) {
-        if (std::abs(reco::deltaPhi(phiOfMuonInteractionRegion(muon), muon.phi())) < 0.5 * M_PI) {
+        if (std::abs(reco::deltaPhi(phiOfMuonInteractionRegion(muon), muon.phi())) < M_PI_2) {
           fillMuonId(iEvent, iSetup, muon, TrackDetectorAssociator::InsideOut);
         } else {
           fillMuonId(iEvent, iSetup, muon, TrackDetectorAssociator::OutsideIn);
@@ -860,7 +862,7 @@ void MuonIdProducer::fillMuonId(edm::Event& iEvent,
   std::vector<reco::MuonChamberMatch> muonChamberMatches;
   unsigned int nubmerOfMatchesAccordingToTrackAssociator = 0;
   for (const auto& chamber : info.chambers) {
-    if (chamber.id.subdetId() == 3 && rpcHitHandle_.isValid())
+    if (chamber.id.subdetId() == MuonSubdetId::RPC && rpcHitHandle_.isValid())
       continue;  // Skip RPC chambers, they are taken care of below)
     reco::MuonChamberMatch matchedChamber;
 
@@ -925,23 +927,29 @@ void MuonIdProducer::fillMuonId(edm::Event& iEvent,
                                      << ", chamber y: " << matchedChamber.y << ", max: " << maxAbsDy_;
       const double matchedSegChDx = std::abs(matchedSegment.x - matchedChamber.x);
       const double matchedSegChDy = std::abs(matchedSegment.y - matchedChamber.y);
-      const double matchedSegChPullX =
-          matchedSegChDx / std::sqrt(std::pow(matchedSegment.xErr, 2) + std::pow(matchedChamber.xErr, 2));
-      const double matchedSegChPullY =
-          matchedSegChDy / std::sqrt(std::pow(matchedSegment.yErr, 2) + std::pow(matchedChamber.yErr, 2));
       if (matchedSegment.xErr > 0 && matchedChamber.xErr > 0)
-        LogTrace("MuonIdentification") << " xpull: " << matchedSegChPullX;
+        LogTrace("MuonIdentification") << " xpull: "
+                                       << matchedSegChDx / std::sqrt(std::pow(matchedSegment.xErr, 2) +
+                                                                     std::pow(matchedChamber.xErr, 2));
       if (matchedSegment.yErr > 0 && matchedChamber.yErr > 0)
-        LogTrace("MuonIdentification") << " ypull: " << matchedSegChPullY;
+        LogTrace("MuonIdentification") << " ypull: "
+                                       << matchedSegChDy / std::sqrt(std::pow(matchedSegment.yErr, 2) +
+                                                                     std::pow(matchedChamber.yErr, 2));
 
       if (matchedSegChDx < maxAbsDx_)
         matchedX = true;
+      else if (matchedSegment.xErr > 0 && matchedChamber.xErr > 0) {
+        const double invMatchedSegChPullX2 = std::pow(matchedSegment.xErr, 2) + std::pow(matchedChamber.xErr, 2);
+        if (matchedSegChDx * matchedSegChDx < maxAbsPullX2_ * invMatchedSegChPullX2)
+          matchedX = true;
+      }
       if (matchedSegChDy < maxAbsDy_)
         matchedY = true;
-      if (matchedSegment.xErr > 0 && matchedChamber.xErr > 0 && matchedSegChPullX < maxAbsPullX_)
-        matchedX = true;
-      if (matchedSegment.yErr > 0 && matchedChamber.yErr > 0 && matchedSegChPullY < maxAbsPullY_)
-        matchedY = true;
+      else if (matchedSegment.yErr > 0 && matchedChamber.yErr > 0) {
+        const double invMatchedSegChPullY2 = std::pow(matchedSegment.yErr, 2) + std::pow(matchedChamber.yErr, 2);
+        if (matchedSegChDy * matchedSegChDy < maxAbsPullY2_ * invMatchedSegChPullY2)
+          matchedY = true;
+      }
       if (matchedX && matchedY) {
         if (matchedChamber.id.subdetId() == MuonSubdetId::ME0)
           matchedChamber.me0Matches.push_back(matchedSegment);
@@ -958,7 +966,7 @@ void MuonIdProducer::fillMuonId(edm::Event& iEvent,
   LogTrace("MuonIdentification") << "RecoMuon/MuonIdProducer :: fillMuonId :: fill RPC info";
   if (rpcHitHandle_.isValid()) {
     for (const auto& chamber : info.chambers) {
-      if (chamber.id.subdetId() != 3)
+      if (chamber.id.subdetId() != MuonSubdetId::RPC)
         continue;  // Consider RPC chambers only
       const auto& lErr = chamber.tState.localError();
       const auto& lPos = chamber.tState.localPosition();

--- a/RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc
+++ b/RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc
@@ -11,6 +11,13 @@
 // user include files
 #include "RecoMuon/MuonIdentification/plugins/MuonIdProducer.h"
 
+#include "DataFormats/Math/interface/deltaPhi.h"
+#include "DataFormats/MuonDetId/interface/MuonSubdetId.h"
+#include "DataFormats/MuonDetId/interface/DTChamberId.h"
+#include "DataFormats/MuonDetId/interface/CSCDetId.h"
+#include "DataFormats/MuonDetId/interface/RPCDetId.h"
+#include "DataFormats/MuonDetId/interface/GEMDetId.h"
+#include "DataFormats/MuonDetId/interface/ME0DetId.h"
 #include "DataFormats/MuonReco/interface/MuonCocktails.h"
 #include "DataFormats/MuonReco/interface/MuonTime.h"
 #include "DataFormats/MuonReco/interface/MuonTimeExtra.h"
@@ -21,15 +28,7 @@
 #include "PhysicsTools/IsolationAlgos/interface/IsoDepositExtractorFactory.h"
 #include "TrackingTools/Records/interface/TrackingComponentsRecord.h"
 
-#include "DataFormats/MuonDetId/interface/MuonSubdetId.h"
-#include "DataFormats/MuonDetId/interface/DTChamberId.h"
-#include "DataFormats/MuonDetId/interface/CSCDetId.h"
-#include "DataFormats/MuonDetId/interface/RPCDetId.h"
-#include "DataFormats/MuonDetId/interface/GEMDetId.h"
-#include "DataFormats/MuonDetId/interface/ME0DetId.h"
-
 #include "RecoMuon/MuonIdentification/interface/MuonMesh.h"
-
 #include "RecoMuon/MuonIdentification/interface/MuonKinkFinder.h"
 
 MuonIdProducer::MuonIdProducer(const edm::ParameterSet& iConfig)
@@ -564,7 +563,8 @@ void MuonIdProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) 
 
         for (auto& muon : *outputMuons) {
           if (muon.innerTrack().get() == trackerMuon.innerTrack().get() &&
-              cos(phiOfMuonInteractionRegion(muon) - phiOfMuonInteractionRegion(trackerMuon)) > 0) {
+              std::abs(reco::deltaPhi(phiOfMuonInteractionRegion(muon), phiOfMuonInteractionRegion(trackerMuon))) <
+                  0.5 * M_PI) {
             newMuon = false;
             muon.setMatches(trackerMuon.matches());
             if (trackerMuon.isTimeValid())
@@ -665,7 +665,7 @@ void MuonIdProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) 
       // predict direction based on the muon interaction region location
       // if it's available
       if (muon.isStandAloneMuon()) {
-        if (cos(phiOfMuonInteractionRegion(muon) - muon.phi()) > 0) {
+        if (std::abs(reco::deltaPhi(phiOfMuonInteractionRegion(muon), muon.phi())) < 0.5 * M_PI) {
           fillMuonId(iEvent, iSetup, muon, TrackDetectorAssociator::InsideOut);
         } else {
           fillMuonId(iEvent, iSetup, muon, TrackDetectorAssociator::OutsideIn);
@@ -925,8 +925,10 @@ void MuonIdProducer::fillMuonId(edm::Event& iEvent,
                                      << ", chamber y: " << matchedChamber.y << ", max: " << maxAbsDy_;
       const double matchedSegChDx = std::abs(matchedSegment.x - matchedChamber.x);
       const double matchedSegChDy = std::abs(matchedSegment.y - matchedChamber.y);
-      const double matchedSegChPullX = matchedSegChDx / std::hypot(matchedSegment.xErr, matchedChamber.xErr);
-      const double matchedSegChPullY = matchedSegChDy / std::hypot(matchedSegment.yErr, matchedChamber.yErr);
+      const double matchedSegChPullX =
+          matchedSegChDx / std::sqrt(std::pow(matchedSegment.xErr, 2) + std::pow(matchedChamber.xErr, 2));
+      const double matchedSegChPullY =
+          matchedSegChDy / std::sqrt(std::pow(matchedSegment.yErr, 2) + std::pow(matchedChamber.yErr, 2));
       if (matchedSegment.xErr > 0 && matchedChamber.xErr > 0)
         LogTrace("MuonIdentification") << " xpull: " << matchedSegChPullX;
       if (matchedSegment.yErr > 0 && matchedChamber.yErr > 0)
@@ -995,7 +997,7 @@ void MuonIdProducer::fillMuonId(edm::Event& iEvent,
         rpcHitMatch.bx = rpcRecHit.BunchX();
 
         const double absDx = std::abs(rpcRecHit.localPosition().x() - chamber.tState.localPosition().x());
-        if (absDx <= 20 or absDx / sqrt(localError.xx()) <= 4)
+        if (absDx <= 20 or absDx * absDx <= 16 * localError.xx())
           matchedChamber.rpcMatches.push_back(rpcHitMatch);
       }
 
@@ -1297,7 +1299,7 @@ void MuonIdProducer::fillMuonIsolation(edm::Event& iEvent,
 }
 
 reco::Muon MuonIdProducer::makeMuon(const reco::Track& track) {
-  const double energy = hypot(track.p(), 0.105658369);
+  const double energy = std::sqrt(track.p() * track.p() + 0.105658369 * 0.105658369);
   const math::XYZTLorentzVector p4(track.px(), track.py(), track.pz(), energy);
   return reco::Muon(track.charge(), p4, track.vertex());
 }

--- a/RecoMuon/MuonIdentification/plugins/MuonIdProducer.h
+++ b/RecoMuon/MuonIdentification/plugins/MuonIdProducer.h
@@ -205,9 +205,9 @@ private:
 
   // matching
   double maxAbsDx_;
-  double maxAbsPullX_;
+  double maxAbsPullX2_;
   double maxAbsDy_;
-  double maxAbsPullY_;
+  double maxAbsPullY2_;
 
   // what information to fill
   bool fillCaloCompatibility_;

